### PR TITLE
RRFS: enable HAILCAST output

### LINKED ISSUE #774: Enable HAILCAST output in RRFS
The HAILCAST hail model is available in FV3...we need to enable HAILCAST output in RRFS.

### PR DESCRIPTION
**Pull request by jaymes-kenyon**
Updates to read-in HAILCAST output from RRFS and subsequently produce GRIB2 output.  This capability existed in the HRRR and is also available in RRFS.  This PR should satisfy issue #774.

### CODE REVIEW

**Comment by jaymes-kenyon** [comment]
This PR is ready for review. The code has been tested successfully in a RRFS case on Jet (CONUS domain).

A sample of netCDF history files containing HAILCAST output available is on Hera:
`/scratch1/BMC/wrfruc/kenyon/upp_development/hailcast`
where the dyn* file contains a 2D array called "**HAILCAST_DHAIL**" (units of mm).  The the hail sizes in this array are very small, but are correctly passed into the GRIB2 files (as "**HAIL**", units of m).

Copying @EricJames-NOAA and @AndrewBenjamin-NOAA for awareness.

File: sorc/ncep_post.fd/INITPOST_NETCDF.f [hunk 1]
```diff
@@ -1056,6 +1057,11 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
       call read_netcdf_2d_para(ncid3d,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
       spval,VarName,rel_vort_maxhy1(ista_2l,jsta_2l))
      if(debugprint)print*,'sample ',VarName,' =',rel_vort_maxhy1(isa,jsa)
+! HAILCAST maximum hail diameter (mm) since last output
+      VarName='HAILCAST_DHAIL'
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f]: @jaymes-kenyon May I ask why this variable name is in upper case? Most of variables from the UFS based models are in lower case.

Comment (jaymes-kenyon) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: Hi @WenMeng-NOAA — I noticed this, too.  Looking in the dyn*.nc files, this seems to be the only array that uses upper case (see screenshot).  I tried  coding `INITPOST_NETCDF.f` with "**hailcast_dhail**" (lower case), but the array was not read-in correctly.
Screenshot 2023-10-25 at 8 34 59 AM

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @jaymes-kenyon Would you check the diag table "diag_table" in the UFS model side if this variable is set to be written out in netcdf file in upper case?

Comment (jaymes-kenyon) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @WenMeng-NOAA — Thanks!  I was able to modify the RRFS diag_table to produce a lower-case array name in the dyn*.nc files.  Updated *.nc files are now available on Hera: 
`/scratch1/BMC/wrfruc/kenyon/upp_development/hailcast/2022072009`
With my latest commit to this PR, I changed `INITPOST_NETCDF.f` to use this lower-case name.  Thanks for the suggestion and the help!

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f]: @jaymes-kenyon Thanks!

**Comment by WenMeng-NOAA** [comment]
@jaymes-kenyon My test with model history files provided by you is at
/scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2022072009
I saw new variable HAIL was generated as:
```
[Wen.Meng@hfe06 post_rrfs_2022072009]$ wgrib2 PRSLEV03.tm00 -match HAIL -stats -s
755:677104823:ndata=1905141:undef=0:mean=3.29272e-09:min=0:max=3.49e-05:cos_wt_mean=3.01373e-09:d=2022072009:HAIL:surface:2-3 hour max fcst:
```
Please let me know if you see issues in my test.

**Comment by jaymes-kenyon** [comment]
@WenMeng-NOAA — This looks good, and the new output is expected.  Thanks for checking!

**Comment by WenMeng-NOAA** [comment]
@FernandoAndrade-NOAA You might refer to my UPP RTs at /scratch1/NCEPDEV/stmp2/Wen.Meng on Hera:
```
-rw-r--r-- 1 Wen.Meng stmp    75 Oct 25 18:42 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
-rw-r--r-- 1 Wen.Meng stmp 23163 Oct 25 18:44 fv3r_2023062800/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 23163 Oct 25 18:44 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 32508 Oct 25 18:44 rtma_2023040400_pe_test/NATLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 32508 Oct 25 18:45 rtma_2023040400/NATLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 24060 Oct 25 18:46 rtma_2023040400_pe_test/PRSLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 32528 Oct 25 18:46 fv3r_2023062800/NATLEV10.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 24060 Oct 25 18:46 rtma_2023040400/PRSLEV00.tm00.diff
-rw-r--r-- 1 Wen.Meng stmp 32528 Oct 25 18:47 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
```

**Comment by FernandoAndrade-NOAA** [comment]
> @FernandoAndrade-NOAA You might refer to my UPP RTs at /scratch1/NCEPDEV/stmp2/Wen.Meng on Hera:
> 
> ```
> -rw-r--r-- 1 Wen.Meng stmp    75 Oct 25 18:42 rap_2020072316_pe_test/WRFPRS.GrbF16.diff
> -rw-r--r-- 1 Wen.Meng stmp 23163 Oct 25 18:44 fv3r_2023062800/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 23163 Oct 25 18:44 fv3r_2023062800_pe_test/PRSLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 32508 Oct 25 18:44 rtma_2023040400_pe_test/NATLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 32508 Oct 25 18:45 rtma_2023040400/NATLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 24060 Oct 25 18:46 rtma_2023040400_pe_test/PRSLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 32528 Oct 25 18:46 fv3r_2023062800/NATLEV10.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 24060 Oct 25 18:46 rtma_2023040400/PRSLEV00.tm00.diff
> -rw-r--r-- 1 Wen.Meng stmp 32528 Oct 25 18:47 fv3r_2023062800_pe_test/NATLEV10.tm00.diff
> ```

My UPP RT results match yours on Hera.

**Comment by WenMeng-NOAA** [comment]
THe UPP RTs were completed on WCOSS2. The baselines of fv3r and 3drtma will be recreated with this PR.

**Comment by FernandoAndrade-NOAA** [comment]
Orion was down for maintenance today apologies for the delay. The Orion UPP RT results seem to match with the Hera changes in fv3r and 3drtma. @WenMeng-NOAA if you'd like to confirm with your own run and compare, mine is over in:
```/work2/noaa/epic/nandoam/regression-testing/upp/orion/808/rundir-upp-ORION```

**Review by FernandoAndrade-NOAA — APPROVED** [review]

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
This PR is ready for merging.

**Pull request merged**